### DISCUSSION TITLE
brand new application, controller not found?

### DISCUSSION THREAD

**Discussion by blm14** [phalcon/cphalcon#15594] [General]
rhel 8.4, php 7.4.6, cphalcon 4.1.2

from phpinfo():

Phalcon is a full stack PHP framework, delivered as a PHP extension, offering lower resource consumption and high performance.
phalcon => enabled
Author => Phalcon Team and contributors
Version => 4.1.2
Build Date => May 23 2021 19:49:51
Powered by Zephir => Version 0.12.21-8a412a1

using phalcon-devtools. from phalcon info:

Environment:
  OS: Linux nine.cpmc.columbia.edu 4.18.0-305.7.1.el8_4.x86_64 #1 SMP Mon Jun 14 17:25:42 EDT 2021 x86_64
  PHP Version: 7.4.6
  PHP SAPI: cli
  PHP Bin: /usr/bin/php
  PHP Extension Dir: /usr/lib64/php/modules
  PHP Bin Dir: /usr/bin
  Loaded PHP config: /etc/php.ini
Versions:
  Phalcon DevTools Version: 4.1.2
  Phalcon Version: 4.1.2
  AdminLTE Version: 3.0.1

Mod rewrite is enabled

from /var/www/html I did this:

sudo phalcon create-project ccadmin --enable-webtools

that worked:

ccadmin/
ccadmin/app
ccadmin/app/views
ccadmin/app/views/index
ccadmin/app/views/index/index.phtml
ccadmin/app/views/layouts
ccadmin/app/views/index.phtml
ccadmin/app/config
ccadmin/app/config/loader.php
ccadmin/app/config/services.php
ccadmin/app/config/router.php
ccadmin/app/config/config.php~
ccadmin/app/config/config.php
ccadmin/app/models
ccadmin/app/controllers
ccadmin/app/controllers/ControllerBase.php
ccadmin/app/controllers/IndexController.php
ccadmin/app/library
ccadmin/app/migrations
ccadmin/cache
ccadmin/public
ccadmin/public/img
ccadmin/public/css
ccadmin/public/temp
ccadmin/public/files
ccadmin/public/js
ccadmin/public/index.php
ccadmin/public/.htaccess
ccadmin/public/webtools.php
ccadmin/public/webtools.config.php
ccadmin/.phalcon
ccadmin/.htaccess
ccadmin/index.html
ccadmin/.htrouter.php

ccadmin/.htrouter.php contents:

<?php

/**
 * This file is part of the Phalcon Developer Tools.
 *
 * (c) Phalcon Team <team@phalcon.io>
 *
 * For the full copyright and license information, please view
 * the LICENSE file that was distributed with this source code.
 */

$uri = urldecode(parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH));

if ($uri !== '/' && file_exists(__DIR__ . '/public' . $uri)) {
    return false;
}

$_GET['_url'] = $_SERVER['REQUEST_URI'];

require_once __DIR__ . '/public/index.php';

attempting to go to server/ccadmin I get this:

CcadminController handler class cannot be loaded
#0 [internal function]: Phalcon\Mvc\Dispatcher->throwDispatchException('CcadminControll...', 2)
#1 [internal function]: Phalcon\Dispatcher\AbstractDispatcher->dispatch()
#2 /var/www/html/ccadmin/public/index.php(43): Phalcon\Mvc\Application->handle('/ccadmin/')
#3 {main}

Why is it not looking for IndexController?

**Comment by blm14** [comment]
CcadminController handler class cannot be loaded
#0 [internal function]: Phalcon\Mvc\Dispatcher->throwDispatchException('CcadminControll...', 2)
#1 [internal function]: Phalcon\Dispatcher\AbstractDispatcher->dispatch()
#2 /var/www/html/ccadmin/public/index.php(43): Phalcon\Mvc\Application->handle('/ccadmin/')
#3 {main}

Formatting messed up. That's what I see when I go to server/ccadmin

Why is it not looking for IndexController?

**Reply by quasipickle** [reply]
The path you pass to `Phalcon\Mvc\Application::handle()` needs to be the path relative to the _application_ root, not the _web_ root.

Since you're passing "/ccadmin/" to `handle()` it thinks "ccadmin" is the name of the controller.  So, Phalcon does its magic and tries to find the `CcadminController` file - like it says on line 0.

Passing "/" to `handle()` should make it look for the IndexController.

However, the URL you pass to handle() is going to need to be generated dynamically.  The docs suggest using $_SERVER["REQUEST_URI"], but that too will include "/ccadmin/".  This should work:

```php
substr($_SERVER["REQUEST_URI"],strlen("/" . basename(realpath(__DIR__."/../"))))
```

This basically passes the part of `$_SERVER["REQUEST_URI"]` that comes after the name of the parent directory.

**Reply by blm14** [reply]
So the thing is I need this application to live inside its own directory in /var/www/html. So I want it to be inside "ccadmin"

That code snippet you posted - is that something I should be changing somewhere? Thanks a ton for the reply but I'm still not sure what to do :-/

**Reply by quasipickle** [reply]
A Phalcon application can easily live inside a subdirectory.  That snippet just removes the directory's part of the path so Phalcon doesn't try to parse it.

To use that snippet, look to line 2 of the error you're getting.  Rather than passing "/ccadmin" to `handle()`, pass the result of that snippet.

**Reply by blm14** [reply]
I changed this:

$router->handle($_SERVER['REQUEST_URI']);

to this:

$router->handle(substr($_SERVER["REQUEST_URI"],strlen("/" . basename(realpath(_\
_DIR__."/../"))))); 

in app/config/router.php

and I am still getting the same error :(

**Reply by quasipickle** [reply]
Try passing "/" to `handle()`

**Reply by blm14** [reply]
contents of router.php are now:

<?php

$router = $di->getRouter();

// Define your routes here

//$router->handle($_SERVER['REQUEST_URI']);

$router->handle('/');

Same error :(

**Comment by blm14** [comment]
kan haz help? :)

**Reply by BeMySlaveDarlin** [reply]
It its your webserver configuration issue.
Looks like vhosts root path is looking on `/var/www/html/` but must `/var/www/html/public`
Can you provide this project webserver vhosts config?

**Reply by blm14** [reply]
from /etc/httpd/conf.d/ssl.conf

<VirtualHost *:80>
   ServerName nine.cpmc.columbia.edu
   Redirect / https://nine.cpmc.columbia.edu/
</VirtualHost>


<VirtualHost _default_:443>

If that's not what you need which config file contains the relevant information?

**Reply by blm14** [reply]
ping! All my non-phalcon applications are working on this server. I do not think it is a vhost config issue but happy to be proven wrong!

**Reply by BeMySlaveDarlin** [reply]
Can you dump what is in `$_SERVER['REQUEST_URI']` ?
The `REQUEST_URI` param in `$_SERVER` superglobal is filled by nginx/apache webserver during request processing.

**Reply by BeMySlaveDarlin** [reply]
Here example of vhost for nginx, N.B for #comments
```
server {
    listen 80;
    listen [::]:80;

    server_name _;

    # Absolute path to public folder in project
    root /app/public;
    index index.php index.html index.htm;

    # Processing uri params to index.php as REQUEST_URI
    location / {
        try_files $uri $uri/ /index.php?_url=$uri&$args;
    }

    location ~ \.php$ {
        fastcgi_pass service_php_fpm:8000;
        include fastcgi-php.conf;

        fastcgi_buffers 16 32k;
        fastcgi_buffer_size 32k;

        fastcgi_connect_timeout 600;
        fastcgi_send_timeout 600;
        fastcgi_read_timeout 600;
    }

    location ~ /\. {
        deny all;
        return 404;
    }
}
```

**Reply by blm14** [reply]
print_r($_SERVER) give me this:



Array
--
  | (
  | [USER] => apache
  | [HOME] => /usr/share/httpd
  | [ORACLE_HOME] => /usr/lib/oracle/19.10/client64/lib
  | [SCRIPT_NAME] => /temp.php
  | [REQUEST_URI] => /temp.php
  | [QUERY_STRING] =>
  | [REQUEST_METHOD] => GET
  | [SERVER_PROTOCOL] => HTTP/1.1
  | [GATEWAY_INTERFACE] => CGI/1.1
  | [REMOTE_PORT] => 55970
  | [SCRIPT_FILENAME] => /var/www/html/temp.php
  | [SERVER_ADMIN] => root@localhost
  | [CONTEXT_DOCUMENT_ROOT] => /var/www/html
  | [CONTEXT_PREFIX] =>
  | [REQUEST_SCHEME] => https
  | [DOCUMENT_ROOT] => /var/www/html
  | [REMOTE_ADDR] => 156.145.28.244
  | [SERVER_PORT] => 443
  | [SERVER_ADDR] => 156.145.15.238
  | [SERVER_NAME] => nine.cpmc.columbia.edu
  | [SERVER_SOFTWARE] => Apache/2.4.37 (Red Hat Enterprise Linux) OpenSSL/1.1.1g
  | [SERVER_SIGNATURE] =>
  | [PATH] => /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
  | [HTTP_SEC_GPC] => 1
  | [HTTP_DNT] => 1
  | [HTTP_COOKIE] => _ga=GA1.2.1976855304.1507053379; _hjid=81d62d00-e0ce-44ab-b2bd-480e08c01329; PHPSESSID=cdc63e4ba3001cadfcac8d88ecc540f7290e1a10b7950956d38105161693c06d; _gcl_au=1.1.1969616417.1626715837; _fbp=fb.1.1626715837302.1629466565; _hjTLDTest=1
  | [HTTP_ACCEPT_LANGUAGE] => en-US,en;q=0.9
  | [HTTP_ACCEPT_ENCODING] => gzip, deflate, br
  | [HTTP_SEC_FETCH_DEST] => document
  | [HTTP_SEC_FETCH_USER] => ?1
  | [HTTP_SEC_FETCH_MODE] => navigate
  | [HTTP_SEC_FETCH_SITE] => none
  | [HTTP_ACCEPT] => text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
  | [HTTP_USER_AGENT] => Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.92 Safari/537.36
  | [HTTP_UPGRADE_INSECURE_REQUESTS] => 1
  | [HTTP_SEC_CH_UA_MOBILE] => ?0
  | [HTTP_SEC_CH_UA] => "Chromium";v="92", " Not A;Brand";v="99", "Google Chrome";v="92"
  | [HTTP_CONNECTION] => keep-alive
  | [HTTP_HOST] => nine.cpmc.columbia.edu
  | [proxy-nokeepalive] => 1
  | [SSL_SESSION_RESUMED] => Resumed
  | [SSL_SESSION_ID] => 1da59762a08bc115f088012f1e037ba571395d75bb92d812eb8a03a6804b540c
  | [SSL_SERVER_A_SIG] => sha256WithRSAEncryption
  | [SSL_SERVER_A_KEY] => rsaEncryption
  | [SSL_SERVER_I_DN] => CN=InCommon RSA Server CA,OU=InCommon,O=Internet2,L=Ann Arbor,ST=MI,C=US
  | [SSL_SERVER_S_DN] => CN=nine.cpmc.columbia.edu,OU=Information Technology,O=Columbia University,street=116th and Broadway,L=New York,ST=New York,postalCode=10027,C=US
  | [SSL_SERVER_V_END] => Feb 22 23:59:59 2022 GMT
  | [SSL_SERVER_V_START] => Feb 22 00:00:00 2021 GMT
  | [SSL_SERVER_M_SERIAL] => B5E1A8BD4F23AB9AA53F43F8C8BC1427
  | [SSL_SERVER_M_VERSION] => 3
  | [SSL_CLIENT_VERIFY] => NONE
  | [SSL_CIPHER_ALGKEYSIZE] => 256
  | [SSL_CIPHER_USEKEYSIZE] => 256
  | [SSL_CIPHER_EXPORT] => false
  | [SSL_CIPHER] => TLS_AES_256_GCM_SHA384
  | [SSL_COMPRESS_METHOD] => NULL
  | [SSL_SECURE_RENEG] => false
  | [SSL_PROTOCOL] => TLSv1.3
  | [SSL_VERSION_LIBRARY] => OpenSSL/1.1.1g
  | [SSL_VERSION_INTERFACE] => mod_ssl/2.4.37
  | [SSL_SERVER_SAN_DNS_0] => nine.cpmc.columbia.edu
  | [SSL_SERVER_I_DN_CN] => InCommon RSA Server CA
  | [SSL_SERVER_I_DN_OU] => InCommon
  | [SSL_SERVER_I_DN_O] => Internet2
  | [SSL_SERVER_I_DN_L] => Ann Arbor
  | [SSL_SERVER_I_DN_ST] => MI
  | [SSL_SERVER_I_DN_C] => US
  | [SSL_SERVER_S_DN_CN] => nine.cpmc.columbia.edu
  | [SSL_SERVER_S_DN_OU] => Information Technology
  | [SSL_SERVER_S_DN_O] => Columbia University
  | [SSL_SERVER_S_DN_L] => New York
  | [SSL_SERVER_S_DN_ST] => New York
  | [SSL_SERVER_S_DN_C] => US
  | [SSL_TLS_SNI] => nine.cpmc.columbia.edu
  | [HTTPS] => on
  | [UNIQUE_ID] => YQvyuqtuuTjMxhr@27ooCQAAAJM
  | [FCGI_ROLE] => RESPONDER
  | [PHP_SELF] => /temp.php
  | [REQUEST_TIME_FLOAT] => 1628172986.8535
  | [REQUEST_TIME] => 1628172986
  | )

**Reply by blm14** [reply]
I'm starting to lose hope here. There has to be a way to solve this!

**Comment by mesaglam** [comment]
Hello, it's 2024 and I'm still struggling with the same error. I've moved my code from one Ubuntu 20.04 system to another 20.04 system. All PHP versions, Phalcon versions, and Nginx versions are the same. However, I'm getting an error that says the indexController is not found. I've even copied the entire /etc/nginx and /etc/php directories to the new system. All configurations are the same, but I'm still having no luck...

**Reply by mesaglam** [reply]
![image](https://github.com/user-attachments/assets/122465b2-3e09-4d9e-bc0a-afb66c591035)
